File: platforms/blackbox/blackbox.py
# Platform implementation: blackbox (Python)
# Thin OS primitives: elapsed time, timers, local key-value store.

import time
import json
import os
import threading

_recording_start = time.monotonic()
_timers = {}
_timer_counter = 0
_timer_lock = threading.Lock()
_store = {}
_store_path = None


def _init_store_path():
    """Set the persistence path next to the running module."""
    global _store_path
    if _store_path is not None:
        return
    import sys
    main = sys.modules.get('__main__')
    if main and hasattr(main, '__file__'):
        _store_path = os.path.join(
            os.path.dirname(os.path.abspath(main.__file__)),
            "blackbox_store.json"
        )
    else:
        _store_path = "blackbox_store.json"
    _load_store()


def _load_store():
    """Load the key-value store from disk."""
    global _store
    if _store_path and os.path.exists(_store_path):
        try:
            with open(_store_path, 'r') as f:
                _store = json.load(f)
        except Exception:
            _store = {}


def _save_store():
    """Save the key-value store to disk."""
    if _store_path is None:
        return
    try:
        with open(_store_path, 'w') as f:
            json.dump(_store, f, indent=2)
    except Exception:
        pass


# timed stream iteration — sleeps dt between values for real-time playback
def _timed_iterate(stream_name, iterator):
    dt = getattr(iterator, 'dt', 0)
    for value in iterator:
        yield value
        if dt and dt > 0:
            time.sleep(dt)


# @zero input number elapsed$
def _get_elapsed() -> float:
    return round((time.monotonic() - _recording_start) * 1000, 1)


# @zero on (string timer) = every (number ms) do (string callback)
def fn_every__number_do__string(ms: float, callback: str) -> str:
    global _timer_counter
    interval = ms / 1000.0

    with _timer_lock:
        _timer_counter += 1
        timer_id = f"timer-{_timer_counter}"

    def tick():
        if timer_id not in _timers:
            return
        _resolve_and_call(callback)
        if timer_id in _timers:
            t = threading.Timer(interval, tick)
            t.daemon = True
            t.start()
            _timers[timer_id] = t

    t = threading.Timer(interval, tick)
    t.daemon = True
    t.start()
    _timers[timer_id] = t
    return timer_id


def _resolve_and_call(callback):
    """Find and call a zero function by name."""
    import sys
    main = sys.modules.get('__main__')
    if main is None:
        return
    fn_name = "fn_" + callback.replace(" ", "_").replace("-", "_")
    fn = getattr(main, fn_name, None)
    if fn and callable(fn):
        try:
            fn()
        except Exception:
            pass


# @zero on cancel timer (string timer)
def fn_cancel_timer__string(timer_id: str):
    t = _timers.pop(timer_id, None)
    if t is not None:
        t.cancel()


# @zero on store locally (string key, string value)
def fn_store_locally__string__string(key: str, value: str):
    _init_store_path()
    _store[key] = value
    _save_store()


# @zero on (string value) = retrieve locally (string key)
def fn_retrieve_locally__string(key: str) -> str:
    _init_store_path()
    return _store.get(key, "")


# @zero on (string result) = stored keys (string prefix)
def fn_stored_keys__string(prefix: str) -> str:
    _init_store_path()
    matches = sorted(k for k in _store if k.startswith(prefix))
    return ",".join(matches)


# @zero on remove locally (string key)
def fn_remove_locally__string(key: str):
    _init_store_path()
    _store.pop(key, None)
    _save_store()


# @zero on inject call (string name) with (string args) result (string result)
def fn_inject_call__string_with__string_result__string(name: str, args: str, result: str):
    """Inject a Call into the runtime input$ stream (flows to action$ via live pipe)."""
    import sys
    main = sys.modules.get('__main__')
    if main is None:
        return
    Call = getattr(main, 'Call', None)
    push = getattr(main, '_push_runtime_input', None)
    if Call and push:
        push(Call(name=name, args=args, result=result))


# @zero on replay with timing [Action actions$]
def fn_replay_with_timing(actions):
    """Replay actions into input$ with original timing gaps."""
    import sys
    main = sys.modules.get('__main__')
    if main is None:
        return
    Call = getattr(main, 'Call', None)
    push = getattr(main, '_push_runtime_input', None)
    if not (Call and push):
        return
    timestamps = getattr(actions, '_timestamps', [])
    for i, action in enumerate(actions):
        if i > 0 and i < len(timestamps) and (i - 1) < len(timestamps):
            delta = min(timestamps[i] - timestamps[i - 1], 1.0)
            if delta > 0:
                time.sleep(delta)
        name = action.get('name', str(action)) if isinstance(action, dict) else getattr(action, 'name', str(action))
        args = action.get('args', '') if isinstance(action, dict) else getattr(action, 'args', '')
        result = action.get('result', '') if isinstance(action, dict) else getattr(action, 'result', '')
        push(Call(name=name, args=args, result=result))

File: platforms/blackbox/blackbox.ts
// Platform implementation: blackbox (TypeScript)
// Thin OS primitives: elapsed time, timers, local key-value store.

const _recording_start: number = performance.now();
const _timers: Map<string, ReturnType<typeof setInterval>> = new Map();
let _timer_counter: number = 0;
const _store: Map<string, string> = new Map();
const _STORAGE_PREFIX: string = "blackbox:";


function _is_browser(): boolean {
    return typeof window !== "undefined" && typeof localStorage !== "undefined";
}


function _load_store(): void {
    if (!_is_browser()) return;
    for (let i = 0; i < localStorage.length; i++) {
        const raw_key = localStorage.key(i);
        if (raw_key && raw_key.startsWith(_STORAGE_PREFIX)) {
            const key = raw_key.slice(_STORAGE_PREFIX.length);
            _store.set(key, localStorage.getItem(raw_key) ?? "");
        }
    }
}


function _save_key(key: string, value: string): void {
    if (_is_browser()) {
        localStorage.setItem(_STORAGE_PREFIX + key, value);
    }
}


function _remove_key(key: string): void {
    if (_is_browser()) {
        localStorage.removeItem(_STORAGE_PREFIX + key);
    }
}


_load_store();


// timed stream iteration — async generator with setTimeout for real-time playback
function _timed_iterate(stream_name: string, iterator: any): any {
    const dt = iterator?.dt ?? 0;
    if (dt > 0) {
        return (async function* () {
            for (const value of iterator) {
                yield value;
                await new Promise(resolve => setTimeout(resolve, dt * 1000));
            }
        })();
    }
    return (function* () {
        for (const value of iterator) {
            yield value;
        }
    })();
}


// @zero input number elapsed$
function _get_elapsed(): number {
    return Math.round((performance.now() - _recording_start) * 10) / 10;
}


// @zero on (string timer) = every (number ms) do (string callback)
function fn_every__number_do__string(ms: number, callback: string): string {
    _timer_counter++;
    const timer_id = `timer-${_timer_counter}`;
    const interval = setInterval(() => {
        _resolve_and_call(callback);
    }, ms);
    _timers.set(timer_id, interval);
    return timer_id;
}


function _resolve_and_call(callback: string): void {
    const fn_name = "fn_" + callback.replace(/ /g, "_").replace(/-/g, "_");
    const fn = (globalThis as any)[fn_name];
    if (typeof fn === "function") {
        try {
            fn();
        } catch (_) {}
    }
}


// @zero on cancel timer (string timer)
function fn_cancel_timer__string(timer_id: string): void {
    const interval = _timers.get(timer_id);
    if (interval !== undefined) {
        clearInterval(interval);
        _timers.delete(timer_id);
    }
}


// @zero on store locally (string key, string value)
function fn_store_locally__string__string(key: string, value: string): void {
    _store.set(key, value);
    _save_key(key, value);
}


// @zero on (string value) = retrieve locally (string key)
function fn_retrieve_locally__string(key: string): string {
    return _store.get(key) ?? "";
}


// @zero on (string result) = stored keys (string prefix)
function fn_stored_keys__string(prefix: string): string {
    const matches: string[] = [];
    for (const k of _store.keys()) {
        if (k.startsWith(prefix)) {
            matches.push(k);
        }
    }
    matches.sort();
    return matches.join(",");
}


// @zero on remove locally (string key)
function fn_remove_locally__string(key: string): void {
    _store.delete(key);
    _remove_key(key);
}
